File: ScalaRewrite/src/main/scala/assignments/musicians/BandSupervisor.scala
package assignments.musicians

import java.util.concurrent.LinkedTransferQueue
import scala.annotation.tailrec
import scala.util.control.Breaks

object BandSupervisor:
  def startBand(maxRetries: Int): Unit =
    val members = collection.mutable.HashMap.empty[String, LinkedTransferQueue[Integer]]
    val queue = LinkedTransferQueue[(String, SkillLevel)]
    members("singer") = addBandMember("singer", SkillLevel.Good, queue)
    members("bass") = addBandMember("bass", SkillLevel.Good, queue)
    members("drum") = addBandMember("drum", SkillLevel.Bad, queue)
    members("guitar") = addBandMember("guitar", SkillLevel.Good, queue)
    loop(maxRetries, queue, members)

  @tailrec
  private def loop(retries: Int, queue: LinkedTransferQueue[(String, SkillLevel)], members: collection.mutable.HashMap[String, LinkedTransferQueue[Integer]]): Unit =
    if retries <= 0 then
      println("The manager is mad and fired the whole band!")
      members.foreach { member => member._2.offer(0) }
      Thread.sleep(3000)
    else
      val (role, skillLevel) = queue.take()
      members.remove(role)
      members(role) = addBandMember(role, skillLevel, queue)
      loop(retries - 1, queue, members)

  @tailrec
  private def playSound(role: String, skillLevel: SkillLevel, musician: Musician, queue: LinkedTransferQueue[(String, SkillLevel)]): Unit =
    Thread.sleep(Musician.delay)
    val playResult = musician.playSound()
    if !playResult then
      queue.offer((role, skillLevel))
    else
      playSound(role, skillLevel, musician, queue)

  private def addBandMember(role: String, skillLevel: SkillLevel, queue: LinkedTransferQueue[(String, SkillLevel)]): LinkedTransferQueue[Integer] =
    val fireQueue = LinkedTransferQueue[Integer]()
    val musician = Musician(role, skillLevel, fireQueue)
    val thread = Thread(() => {
      playSound(role, skillLevel, musician, queue)
    })
    thread.start()
    fireQueue

File: RubyRewrite/lib/assignments/musicians/band_supervisor.rb
# frozen_string_literal: true

require_relative 'musician'

module BandSupervisor
  DELAY = 0.75

  def self.start_band(max_retries)
    queue = Thread::Queue.new
    members = {
      singer: add_band_member('singer', :good, queue),
      bass: add_band_member('bass', :good, queue),
      drum: add_band_member('drum', :bad, queue),
      guitar: add_band_member('guitar', :good, queue)
    }

    loop do
      new_data = queue.pop
      max_retries -= 1
      if max_retries <= 0
        puts 'The manager is mad and fired the whole band!'
        members.each do |_, q|
          q.enq(0)
        end
        sleep 3.0
        break
      end

      members[new_data[:role]] = add_band_member(new_data[:role], new_data[:skill_level], queue)
    end
  end

  def self.add_band_member(role, skill_level, queue)
    fire_queue = Thread::Queue.new
    musician = Musician.new(role, skill_level, fire_queue)
    Thread.new(musician) do |m|
      loop do
        sleep DELAY
        play_result = m.play_sound
        next if play_result

        queue.enq({ role:, skill_level: })
        break
      end
    end
    fire_queue
  end
end
